File: double-pointer/src/q876.py
# Definition for singly-linked list.
import pytest

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        if head.next is None:
            return head

        fp = head
        np = head

        while True:
            np = np.next
            if np is None:
                return fp
            np = np.next
            if np is None:
                return fp.next

            fp = fp.next



def test_mid():
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, None)))))

    assert Solution().middleNode(head).val == 3

    head = ListNode(0, head)
    assert Solution().middleNode(head).val == 3


if __name__ == '__main__':
    pytest.main(["./q876.py", "-s", "-v"])

File: double-pointer/src/q876.rs
use super::*;

struct Solution ();
impl Solution {
    pub fn middle_node(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut fast = head.as_ref().unwrap();
        let mut slow = head.as_ref();

        if fast.next.is_none() { return head; }

        while !fast.next.is_none() && !fast.next.as_ref().unwrap().next.is_none() {
            fast = fast.next.as_ref().unwrap().next.as_ref().unwrap();
            if fast.next.is_none() {
                return slow.unwrap().next.clone();
            }
            slow = slow.unwrap().next.as_ref();

        }
        slow.unwrap().next.clone()


    }
}


#[cfg(test)]
mod tests {
  use super::*;

  #[test] fn test_1() {
      assert_eq!(Solution::middle_node(linklist!(1,2,3,4,5)).unwrap().val, 3);
      assert_eq!(Solution::middle_node(linklist!(1,2,3,4,5, 6)).unwrap().val, 4);


  }
}